src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,377 @@
+        public bool IsEmpty => Length == 0;
+
+        /// <summary>
+        /// Determins if the <see cref="ReadOnlyBuffer{T}"/> is a single <see cref="Memory{Byte}"/>.
+        /// </summary>
+        public bool IsSingleSegment => BufferStart.Segment == BufferEnd.Segment;
+
+        public ReadOnlyMemory<T> First
+        {
+            get
+            {
+                TryGetBuffer(BufferStart, BufferEnd, out var first, out _);
+                return first;
+            }
+        }
+
+        /// <summary>
+        /// A position to the start of the <see cref="ReadOnlyBuffer{T}"/>.
+        /// </summary>
+        public SequencePosition Start => BufferStart;
+
+        /// <summary>
+        /// A position to the end of the <see cref="ReadOnlyBuffer{T}"/>
+        /// </summary>
+        public SequencePosition End => BufferEnd;
+
+        private ReadOnlyBuffer(object startSegment, int startIndex, object endSegment, int endIndex)
+        {
+            Debug.Assert(startSegment != null);
+            Debug.Assert(endSegment != null);
+
+            BufferStart = new SequencePosition(startSegment, startIndex);
+            BufferEnd = new SequencePosition(endSegment, endIndex);
+        }
+
+        public ReadOnlyBuffer(IMemoryList<T> startSegment, int startIndex, IMemoryList<T> endSegment, int endIndex)
+        {
+            Debug.Assert(startSegment != null);
+            Debug.Assert(endSegment != null);
+            Debug.Assert(startSegment.Memory.Length >= startIndex);
+            Debug.Assert(endSegment.Memory.Length >= endIndex);
+
+            BufferStart = new SequencePosition(startSegment, startIndex | MemoryListStartMask);
+            BufferEnd = new SequencePosition(endSegment, endIndex | MemoryListEndMask);
+        }
+
+        public ReadOnlyBuffer(T[] array) : this(array, 0, array.Length)
+        {
+        }
+
+        public ReadOnlyBuffer(T[] array, int offset, int length)
+        {
+            if (array == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            }
+
+            BufferStart = new SequencePosition(array, offset | ArrayStartMask);
+            BufferEnd = new SequencePosition(array, offset + length | ArrayEndMask);
+        }
+
+        // Consumer is expected to manage lifetime of memory until
+        // ReadOnlyBuffer is not used anymore
+        public ReadOnlyBuffer(Memory<T> memory)
+        {
+            var segment = new ReadOnlyBufferSegment
+            {
+                Memory = memory
+            };
+            BufferStart = new SequencePosition(segment, 0 | MemoryListStartMask);
+            BufferEnd = new SequencePosition(segment, memory.Length | MemoryListEndMask);
+        }
+
+        public ReadOnlyBuffer(IEnumerable<Memory<T>> buffers)
+        {
+            ReadOnlyBufferSegment segment = null;
+            ReadOnlyBufferSegment first = null;
+            foreach (var buffer in buffers)
```

Comment:
nit: state type explicitly instead of using `var` since it can't be inferred from this line alone.